rmw_fastrtps_shared_cpp/test/test_topic_cache.cpp

```diff
@@ -0,0 +1,263 @@
+  GUID_t participant_guid[2];
+  GUID_t guid[2];
+  void SetUp()
+  {
+    // Create instance handlers
+    for (int i = 0; i < 2; i++) {
+      guid[i] = GUID_t(GuidPrefix_t(), i + 100);
+      participant_guid[i] = GUID_t(GuidPrefix_t(), i + 1);
+      participant_instance_handler[i] = participant_guid[i];
+    }
+
+    // Populating WriterQos -> which is from the DDS layer and
+    // rmw_qos_profile_t which is from rmw/types.h.
+    // This is done to test if topic_cache.getTopicNameToTopicData() returns
+    // the correct value in rmw_qos_profile_t for a given WriterQos
+
+    // DDS qos
+    qos[0].m_durability.kind = eprosima::fastrtps::TRANSIENT_DURABILITY_QOS;
+    qos[0].m_reliability.kind = eprosima::fastrtps::RELIABLE_RELIABILITY_QOS;
+    qos[0].m_deadline.period = {123, 5678};
+    qos[0].m_lifespan.duration = {190, 1234};
+    qos[0].m_liveliness.kind = eprosima::fastrtps::MANUAL_BY_TOPIC_LIVELINESS_QOS;
+    qos[0].m_liveliness.lease_duration = {80, 5555555};
+
+    // equivalent rmq qos
+    rmw_qos[0].durability = RMW_QOS_POLICY_DURABILITY_UNKNOWN;
+    rmw_qos[0].reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+    rmw_qos[0].liveliness = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC;
+    rmw_qos[0].liveliness_lease_duration.sec = 80u;
+    rmw_qos[0].liveliness_lease_duration.nsec = 5555555u;
+    rmw_qos[0].deadline.sec = 123u;
+    rmw_qos[0].deadline.nsec = 5678u;
+    rmw_qos[0].lifespan.sec = 190u;
+    rmw_qos[0].lifespan.nsec = 1234u;
+
+    // DDS qos
+    qos[1].m_durability.kind = eprosima::fastrtps::TRANSIENT_DURABILITY_QOS;
+    qos[1].m_reliability.kind = eprosima::fastrtps::RELIABLE_RELIABILITY_QOS;
+    qos[1].m_deadline.period = {12, 1234};
+    qos[1].m_lifespan.duration = {19, 5432};
+    qos[1].m_liveliness.kind = eprosima::fastrtps::MANUAL_BY_TOPIC_LIVELINESS_QOS;
+    qos[1].m_liveliness.lease_duration = {8, 78901234};
+
+    // equivalent rmq qos
+    rmw_qos[1].durability = RMW_QOS_POLICY_DURABILITY_UNKNOWN;
+    rmw_qos[1].reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+    rmw_qos[1].liveliness = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC;
+    rmw_qos[1].liveliness_lease_duration.sec = 8u;
+    rmw_qos[1].liveliness_lease_duration.nsec = 78901234u;
+    rmw_qos[1].deadline.sec = 12u;
+    rmw_qos[1].deadline.nsec = 1234u;
+    rmw_qos[1].lifespan.sec = 19u;
+    rmw_qos[1].lifespan.nsec = 5432u;
+
+    // Add data to topic_cache
+    topic_cache.addTopic(participant_instance_handler[0], guid[0], "topic1", "type1", qos[0]);
+    topic_cache.addTopic(participant_instance_handler[0], guid[0], "topic2", "type2", qos[0]);
+    topic_cache.addTopic(participant_instance_handler[1], guid[1], "topic1", "type1", qos[1]);
+    topic_cache.addTopic(participant_instance_handler[1], guid[1], "topic2", "type1", qos[1]);
+  }
+};
+
+TEST_F(TopicCacheTestFixture, test_topic_cache_get_topic_types)
+{
+  const auto & topic_type_map = this->topic_cache.getTopicToTypes();
```

Comment:
@jaisontj if `getTopicToTypes()` returns an rvalue, `topic_type_map` shouldn't be a reference.

Reply:
This just can't be a non-`const` reference; a `const` reference actually will extend a temporary object's lifetime.

But let's make it a value instead of a reference anyway.